src/ziterate.c

```diff
@@ -0,0 +1,207 @@
+/*
+ * ZMap Copyright 2013 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+/*
+ * ZIterate is a simple utility that will iteratate over the IPv4
+ * space in a pseudo-random fashion, utilizing the sharding capabilities
+ * of zmap to enable this iteration to be split among multiple instances of
+ * ziterate.
+ */
+
+#define _GNU_SOURCE
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <getopt.h>
+#include <assert.h>
+#include <unistd.h>
+
+#include "iterator.h"
+#include "../lib/includes.h"
+#include "../lib/blacklist.h"
+#include "../lib/logger.h"
+#include "../lib/random.h"
+#include "state.h"
+#include "validate.h"
+#include "zitopt.h"
+
+struct zit_conf {
+	char *blacklist_filename;
+	char *whitelist_filename;
+	char *log_filename;
+	int check_duplicates;
+	int ignore_errors;
+	int verbosity;
+	// sharding options
+	uint8_t shard_num;
+	uint8_t total_shards;
+	uint64_t seed;
+	aesrand_t *aes;
+};
+
+#define SET_BOOL(DST,ARG) \
+{ if (args.ARG##_given) { (DST) = 1; }; }
+
+int main(int argc, char **argv)
+{
+	struct zit_conf conf;
+
+	memset(&conf, 0, sizeof(struct zit_conf));
+	conf.verbosity = 3;
+	int no_dupchk_pres = 0;
+	conf.ignore_errors = 0;
+
+	struct gengetopt_args_info args;
+	struct cmdline_parser_params *params;
+	params = cmdline_parser_params_create();
+	assert(params);
+	params->initialize = 1;
+	params->override = 0;
+	params->check_required = 0;
+
+	if (cmdline_parser_ext(argc, argv, &args, params) != 0) {
+		exit(EXIT_SUCCESS);
+	}
+
+	// Handle help text and version
+	if (args.help_given) {
+		cmdline_parser_print_help();
+		exit(EXIT_SUCCESS);
+	}
+	if (args.version_given) {
+		cmdline_parser_print_version();
+		exit(EXIT_SUCCESS);
+	}
+
+	// Set the log file and metadata file
+	if (args.log_file_given) {
+		conf.log_filename = strdup(args.log_file_arg);
+	}
+	if (args.verbosity_given) {
+		conf.verbosity = args.verbosity_arg;
+	}
+
+	// Blacklist and whitelist
+	if (args.blacklist_file_given) {
+		conf.blacklist_filename = strdup(args.blacklist_file_arg);
+	}
+	if (args.whitelist_file_given) {
+		conf.whitelist_filename = strdup(args.whitelist_file_arg);
+	}
+
+	// Read the boolean flags
+	SET_BOOL(no_dupchk_pres, no_duplicate_checking);
+	conf.check_duplicates = !no_dupchk_pres;
```

Comment:
My thought was it would take a whitelist (e.g. an Alexa lookup), and shuffle the (possibly duplicate) IPs, but in general not sure `--no-duplicate-checking` makes sense here. Removing it is probably the right call.